{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running Bash in Jupyter notebook\n",
    "To run Bash commands in a Jupyter notebook, you need to install the Bash kernel for Jupyter. For that, run the following commands in the terminal:\n",
    "```\n",
    "pip install bash_kernel\n",
    "python -m bash_kernel.install\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test a bash command\n",
    "ls -lh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 1: use fastq-dump to download from GEO database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check fastq-dump options\n",
    "fastq-dump --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To download the data from SRR3649298 run\n",
    "```\n",
    "fastq-dump --accession SRR3649298 --split-files --outdir fastq --gzip\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download reads from SRR3649298\n",
    "fastq-dump --accession SRR3649298 --split-files --outdir fastq --gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 2: download data from ENA (European Nucleotide Archive)\n",
    "The same data (SRR3649298) is available here, in the fastq format:\n",
    "https://www.ebi.ac.uk/ena/data/view/SRR3649298  \n",
    "The raw fastq files are available from these links:  \n",
    "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_1.fastq.gz  \n",
    "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_2.fastq.gz  \n",
    "\n",
    "To download these files, run:\n",
    "```\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_1.fastq.gz -P fastq\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_2.fastq.gz -P fastq\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_1.fastq.gz -P fastq\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_2.fastq.gz -P fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the size of the downloaded files\n",
    "ls -lh fastq/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the format of the fastq files\n",
    "gunzip -c fastq/SRR3649298_1.fastq.gz | head -n 12 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract 100,000 reads (400,000 lines).\n",
    "gunzip -c fastq/SRR3649298_1.fastq.gz | head -n 400000 > fastq/test_R1.fastq\n",
    "gunzip -c fastq/SRR3649298_2.fastq.gz | head -n 400000 > fastq/test_R2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the new files\n",
    "ls -lh fastq/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read alignment (Bowtie2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the options of Bowtie2\n",
    "bowtie2 --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many Bowtie2 index can be downloaded directly from Illumina\n",
    "https://support.illumina.com/sequencing/sequencing_software/igenome.html\n",
    "\n",
    "The sacCer3 archive is available at\n",
    "ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Saccharomyces_cerevisiae/UCSC/sacCer3/Saccharomyces_cerevisiae_UCSC_sacCer3.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the Illumina archive with the sacCer3 genome\n",
    "wget ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Saccharomyces_cerevisiae/UCSC/sacCer3/Saccharomyces_cerevisiae_UCSC_sacCer3.tar.gz\n",
    "\n",
    "# Unpack the archive\n",
    "tar -xzf Saccharomyces_cerevisiae_UCSC_sacCer3.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the downloaded files\n",
    "ls -lh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# View the directory tree for the downloaded folder\n",
    "tree -d Saccharomyces_cerevisiae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the sacCer3 index to a new folder, e.g. bt2_index_sacCer3\n",
    "mkdir bt2_index_sacCer3\n",
    "cp Saccharomyces_cerevisiae/UCSC/sacCer3/Sequence/Bowtie2Index/* bt2_index_sacCer3/\n",
    "ls -lh bt2_index_sacCer3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, one can use `bowtie2-build` to build the index from the fasta file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the options\n",
    "bowtie2-build --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a folder for the aligned data (sam/bam files)\n",
    "mkdir -p bam\n",
    "\n",
    "# Align reads with bowtie2\n",
    "bowtie2 -x bt2_index_sacCer3/genome -1 fastq/test_R1.fastq -2 fastq/test_R2.fastq -S bam/test.sam 2> bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the size of the aligned reads (sam file) \n",
    "ls -lh bam/test.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# View the alignment report generated by bowtie2\n",
    "cat bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are some reads that could not be aligned as a proper concordant pair, or did not align at all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check bowtie2 options for suppressing bad alignments\n",
    "bowtie2 --help | grep \"suppress\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Align reads with bowtie2, discarding unpaired reads, and reads that are aligned discordantly, and unaligned reads \n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -x bt2_index_sacCer3/genome -1 fastq/test_R1.fastq -2 fastq/test_R2.fastq -S bam/test.sam 2> bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the first lines of the sam file\n",
    "head -30 bam/test.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Faster way of using Bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the number of CPU cores\n",
    "CORES=$(getconf _NPROCESSORS_ONLN)\n",
    "echo $CORES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run bowtie2 on multiple cores to improve speed\n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -p $((CORES-1)) \\\n",
    "        -x bt2_index_sacCer3/genome -1 fastq/test_R1.fastq -2 fastq/test_R2.fastq -S bam/test.sam 2> bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, that was much faster. Now let's store the alignments in a more compact way (bam file instead of sam file). We can convert the sam to bam format using `samtools view`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the arguments of samtools view\n",
    "samtools view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the sam file to bam format\n",
    "samtools view -b --threads $((CORES-1)) bam/test.sam > bam/test.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's compare the sizes of sam and bam files\n",
    "ls -lh bam/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The bam file is about 5 times smaller than the corresponding sam file. Bam is a binary file format so trying to read it as a regular text won't work.\n",
    "```\n",
    "head bam/test.bam\n",
    "```\n",
    "will output some nonsense. The proper way of viewing a bam file is using `samtools view`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Proper way of listing the alignments from a bam file\n",
    "samtools view bam/test.bam | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the total number of alignments\n",
    "samtools view bam/test.bam | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 183,030 reads, corresponding to 91,515 DNA fragments (paired-end reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's look again at the bowtie2 log:\n",
    "cat bam/test.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These 91,515 properly aligned pairs are the sum of: 76,533 + 14,982"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore the bam file in more detail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Explore the columns of the bam file\n",
    "# Column 1: alignment name\n",
    "samtools view bam/test.bam | head | cut -f 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 2: Sum of multiple bitwise flags\n",
    "# See https://samtools.github.io/hts-specs/SAMv1.pdf for an explanation of different FLAGs\n",
    "samtools view bam/test.bam | head | cut -f 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One can easily check the meaning of each FLAG using the following website:\n",
    "https://broadinstitute.github.io/picard/explain-flags.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 3: chromosome name\n",
    "samtools view bam/test.bam | head | cut -f 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 4: 1-based leftmost mapping position\n",
    "samtools view bam/test.bam | head | cut -f 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 5: mapping quality\n",
    "# −10 log10 Prob(mapping position is wrong)\n",
    "samtools view bam/test.bam | head | cut -f 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 6: CIGAR string\n",
    "# Column 7: ‘=’ if the paired read was aligned on the same chromosome\n",
    "# Column 8: position of the paired read\n",
    "# Column 9: length of the whole DNA fragment whose ends were sequenced (+/- for left/right end)\n",
    "samtools view bam/test.bam | head | cut -f 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 10: DNA sequence\n",
    "samtools view bam/test.bam | head | cut -f 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Column 11: ASCII representation of base quality plus 33\n",
    "samtools view bam/test.bam | head | cut -f 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# By dafault the sam/bam files are sorted by the first column (read name)\n",
    "# Show the following 3 columns: 1 - read name, 3 - chromosome, 4 - position)\n",
    "samtools view bam/test.bam | head | cut -f 1,3,4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Many tools require the bam file to be sorted such that the alignments occur in “genome order”. \n",
    "# That is, ordered positionally based upon their alignment coordinates on each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We can re-sort the BAM file using \n",
    "samtools sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools sort -o bam/test.sorted.bam --threads $((CORES-1)) bam/test.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the sorted bam file (columns 1 - read name, 3 - chromosome, 4 - position)\n",
    "samtools view bam/test.sorted.bam | head | cut -f 1,3,4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Indexing a position sorted bam file allows one to quickly extract alignments overlapping particular genomic regions.\n",
    "# This indexing is done using\n",
    "samtools index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools index -@ $((CORES-1)) bam/test.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the sizes (the index file .bai is very small)\n",
    "ls -lh bam/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pipes\n",
    "All the above operations could be run in a more compact way, using pipes (see https://www.geeksforgeeks.org/piping-in-unix-or-linux/ for more info on the pipe operator)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pipes to perform all these operations in one step\n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -p $((CORES-1)) \\\n",
    "        -x bt2_index_sacCer3/genome -1 fastq/test_R1.fastq -2 fastq/test_R2.fastq 2> bam/test2.bt2.log \\\n",
    "        | samtools view -b -@ $((CORES-1)) - \\\n",
    "        | samtools sort -o bam/test2.sorted.bam -@ $((CORES-1)) -\n",
    "samtools index -@ $((CORES-1)) bam/test2.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls -lh bam/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll switch now to another Jupyter notebook and analyze the aligned data in R.\n",
    "\n",
    "Further reading resources: <br>\n",
    "https://samtools.github.io/hts-specs/SAMv1.pdf <br>\n",
    "http://quinlanlab.org/tutorials/samtools/samtools.html <br>\n",
    "http://biobits.org/samtools_primer.html <br>\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
