{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intro to basic command line tools (samtools, bowtie2, etc.)\n",
    "### Running Bash commands inside a Jupyter notebook\n",
    "To run Bash commands in a Jupyter notebook, you need to install the Bash kernel for Jupyter. For that, run the following commands in the terminal:\n",
    "```\n",
    "pip install bash_kernel\n",
    "python -m bash_kernel.install\n",
    "```\n",
    "If the Bash kernel is installed and selected as the kernel for the current Jupyter notebook, we can run Bash commands inside the notebook. For example, we can use the `tree` command to show the contents of the current folder. Please make sure that you have downloaded all of the following files, and the tree of your current folder looks like this: \n",
    "```\n",
    ".\n",
    "├── Part_1_MNase-seq_intro.pptx\n",
    "├── Part_2_Bash_commands.ipynb\n",
    "├── Part_3_R_commands.ipynb\n",
    "├── data_backup\n",
    "│   ├── SRR3649298.bt2.log\n",
    "│   ├── SRR3649298.sorted.bam\n",
    "│   ├── SRR3649298.sorted.bam.bai\n",
    "│   ├── SRR3649298_1.fastq.gz\n",
    "│   ├── SRR3649298_2.fastq.gz\n",
    "│   └── TxDbFromUCSC.sqlite\n",
    "├── info.txt\n",
    "└── sacCer3_genome\n",
    "    └── genome.fa\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".\n",
      "├── Part_1_MNase-seq_intro.pptx\n",
      "├── Part_2_Bash_commands.ipynb\n",
      "├── Part_3_R_commands.ipynb\n",
      "├── data_backup\n",
      "│   ├── SRR3649298.bt2.log\n",
      "│   ├── SRR3649298.sorted.bam\n",
      "│   ├── SRR3649298.sorted.bam.bai\n",
      "│   ├── SRR3649298_1.fastq.gz\n",
      "│   ├── SRR3649298_2.fastq.gz\n",
      "│   └── TxDbFromUCSC.sqlite\n",
      "├── info.txt\n",
      "└── sacCer3_genome\n",
      "    └── genome.fa\n",
      "\n",
      "2 directories, 11 files\n"
     ]
    }
   ],
   "source": [
    "tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sequencing data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Option 1: GEO database / Sequence Read Archive (SRA)  \n",
    "https://www.ncbi.nlm.nih.gov/geo/  \n",
    "https://www.ncbi.nlm.nih.gov/sra  \n",
    "Use `fastq-dump` to download sequencing data from GEO database or Sequence Read Archive (SRA). To download the MNase-seq data with the accession number SRR3649298, run\n",
    "```\n",
    "fastq-dump --accession SRR3649298 --split-files --outdir fastq_files --gzip\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Option 2: European Nucleotide Archive (ENA)  \n",
    "https://www.ebi.ac.uk/ena  \n",
    "Use the direct links to the fastq files provided on their website. The same data (SRR3649298) is available on ENA in different formats:\n",
    "https://www.ebi.ac.uk/ena/data/view/SRR3649298  \n",
    "\n",
    "The raw fastq files are available from these links:  \n",
    "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_1.fastq.gz  \n",
    "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_2.fastq.gz  \n",
    "\n",
    "To download these files, run:\n",
    "```\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_1.fastq.gz -P fastq\n",
    "wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR364/008/SRR3649298/SRR3649298_2.fastq.gz -P fastq\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These files can be very big (100 MB - 1 TB in size). To prevent big downloads, I have copied the necessary `fastq` files into the folder `data_backup`. Let's assume we have downloaded these files using the above commands (I will just copy the files from the local folder)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p fastq_files\n",
    "cp data_backup/*.fastq.gz fastq_files/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 1387616\n",
      "-rw-r--r--@ 1 cherejirv  1360859114   334M Nov 22 16:07 SRR3649298_1.fastq.gz\n",
      "-rw-r--r--@ 1 cherejirv  1360859114   343M Nov 22 16:07 SRR3649298_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "# Check the size of the downloaded files\n",
    "ls -lh fastq_files/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that these are zipped files, and they occupy ~700 MB of space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3649298.1 HWI-D00638:56:C6FTLANXX:2:1101:17378:1996/1\n",
      "NTAAACTGTGCTCGGCCAGTGGAGACGTACAAATTCCACTACTATTCGTA\n",
      "+\n",
      "#<BBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "@SRR3649298.2 HWI-D00638:56:C6FTLANXX:2:1101:18353:1996/1\n",
      "NGTTCCCTGTGTTCTAAAAATTGGAAAAACATGGCTATTAAAACCTGTGG\n",
      "+\n",
      "#<BBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "@SRR3649298.3 HWI-D00638:56:C6FTLANXX:2:1101:1209:2084/1\n",
      "NTCTCAAATCATTTAATAACTACAAAGAAACAAAAAAAAATCAATGTTGA\n",
      "+\n",
      "#<<BBFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFFFFFF\n"
     ]
    }
   ],
   "source": [
    "# Check the format of the fastq files. Print the first 12 lines of the first file\n",
    "gunzip -c fastq_files/SRR3649298_1.fastq.gz | head -n 12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's work first only with a small subset of 100,000 reads. Since each short read occupies 4 lines in the fastq files, we need to copy the first 400,000 lines from each `fastq` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract 100,000 reads (400,000 lines) from the original fastq files\n",
    "gunzip -c fastq_files/SRR3649298_1.fastq.gz | head -n 400000 > fastq_files/test_R1.fastq\n",
    "gunzip -c fastq_files/SRR3649298_2.fastq.gz | head -n 400000 > fastq_files/test_R2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 1452608\n",
      "-rw-r--r--@ 1 cherejirv  1360859114   334M Nov 22 16:07 SRR3649298_1.fastq.gz\n",
      "-rw-r--r--@ 1 cherejirv  1360859114   343M Nov 22 16:07 SRR3649298_2.fastq.gz\n",
      "-rw-r--r--  1 cherejirv  1360859114    16M Nov 22 16:07 test_R1.fastq\n",
      "-rw-r--r--  1 cherejirv  1360859114    16M Nov 22 16:07 test_R2.fastq\n"
     ]
    }
   ],
   "source": [
    "# Check the new files\n",
    "ls -lh fastq_files/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignment of paired-end reads (Bowtie2)\n",
    "The reads from the fastq files were obtained by digesting yeast DNA using microccocal nuclease (MNase). Let's detect the genomic loci where these DNA fragments originated from. For this we will align the reads to the yeast genome, using the Bowtie2 aligner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.3.4.3 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\n",
      "Usage: \n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r> | --interleaved <i>} [-S <sam>]\n",
      "\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <r>        Files with unpaired reads.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <i>        Files with interleaved paired-end FASTQ reads\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <sam>      File for SAM output (default: stdout)\n",
      "\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\n",
      "\n",
      "Options (defaults in parentheses):\n",
      "\n",
      " Input:\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\n",
      "  --tab5             query input files are TAB5 .tab5\n",
      "  --tab6             query input files are TAB6 .tab6\n",
      "  --qseq             query input files are in Illumina's qseq format\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\n",
      "  -r                 query input files are raw one-sequence-per-line\n",
      "  -F k:<int>,i:<int> query input files are continuous FASTA where reads\n",
      "                     are substrings (k-mers) extracted from a FASTA file <s>\n",
      "                     and aligned at offsets 1, 1+i, 1+2i ... end of reference\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\n",
      "  --trim-to [3:|5:]<int> trim reads exceeding <int> bases from either 3' or 5' end\n",
      "                     If the read end is not specified then it defaults to 3 (0)\n",
      "  --phred33          qualities are Phred+33 (default)\n",
      "  --phred64          qualities are Phred+64\n",
      "  --int-quals        qualities encoded as space-delimited integers\n",
      "\n",
      " Presets:                 Same as:\n",
      "  For --end-to-end:\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\n",
      "\n",
      "  For --local:\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\n",
      "\n",
      " Alignment:\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\n",
      "  --nofw             do not align forward (original) version of read (off)\n",
      "  --norc             do not align reverse-complement version of read (off)\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\n",
      "                     scan for the optimal seeded alignments\n",
      "  --end-to-end       entire read must align; no clipping (on)\n",
      "   OR\n",
      "  --local            local alignment; ends might be soft clipped (off)\n",
      "\n",
      " Scoring:\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\n",
      "\n",
      " Reporting:\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\n",
      "   OR\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\n",
      "   OR\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\n",
      "\n",
      " Effort:\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\n",
      "\n",
      " Paired-end:\n",
      "  -I/--minins <int>  minimum fragment length (0)\n",
      "  -X/--maxins <int>  maximum fragment length (500)\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\n",
      "  --no-discordant    suppress discordant alignments for paired reads\n",
      "  --dovetail         concordant when mates extend past each other\n",
      "  --no-contain       not concordant when one mate alignment contains other\n",
      "  --no-overlap       not concordant when mates overlap at all\n",
      "\n",
      " Output:\n",
      "  -t/--time          print wall-clock time taken by search phases\n",
      "  --un <path>        write unpaired reads that didn't align to <path>\n",
      "  --al <path>        write unpaired reads that aligned at least once to <path>\n",
      "  --un-conc <path>   write pairs that didn't align concordantly to <path>\n",
      "  --al-conc <path>   write pairs that aligned concordantly at least once to <path>\n",
      "    (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\n",
      "    --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\n",
      "  --quiet            print nothing to stderr except serious errors\n",
      "  --met-file <path>  send metrics to file at <path> (off)\n",
      "  --met-stderr       send metrics to stderr (off)\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\n",
      "  --no-unal          suppress SAM records for unaligned reads\n",
      "  --no-head          suppress header lines, i.e. lines starting with @\n",
      "  --no-sq            suppress @SQ header lines\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\n",
      "                     Note: @RG line only printed when --rg-id is set.\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\n",
      "  --sam-no-qname-trunc Suppress standard behavior of truncating readname at first whitespace \n",
      "                      at the expense of generating non-standard SAM.\n",
      "  --xeq              Use '='/'X', instead of 'M,' to specify matches/mismatches in SAM record.\n",
      "  --soft-clipped-unmapped-tlen Exclude soft-clipped bases when reporting TLEN\n",
      "\n",
      " Performance:\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\n",
      "  --reorder          force SAM output order to match order of input reads\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\n",
      "\n",
      " Other:\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\n",
      "  --seed <int>       seed for random number generator (0)\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\n",
      "  --version          print version information and quit\n",
      "  -h/--help          print this usage message\n"
     ]
    }
   ],
   "source": [
    "# Check the options of Bowtie2\n",
    "bowtie2 --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll need to create a Bowtie2 index corresponding to the yeast genome. For this we'll use `bowtie2-build`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.3.4.3 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\n",
      "Usage: bowtie2-build [options]* <reference_in> <bt2_index_base>\n",
      "    reference_in            comma-separated list of files with ref sequences\n",
      "    bt2_index_base          write bt2 data to files with this dir/basename\n",
      "*** Bowtie 2 indexes work only with v2 (not v1).  Likewise for v1 indexes. ***\n",
      "Options:\n",
      "    -f                      reference files are Fasta (default)\n",
      "    -c                      reference sequences given on cmd line (as\n",
      "                            <reference_in>)\n",
      "    --large-index           force generated index to be 'large', even if ref\n",
      "                            has fewer than 4 billion nucleotides\n",
      "    --debug                 use the debug binary; slower, assertions enabled\n",
      "    --sanitized             use sanitized binary; slower, uses ASan and/or UBSan\n",
      "    --verbose               log the issued command\n",
      "    -a/--noauto             disable automatic -p/--bmax/--dcv memory-fitting\n",
      "    -p/--packed             use packed strings internally; slower, less memory\n",
      "    --bmax <int>            max bucket sz for blockwise suffix-array builder\n",
      "    --bmaxdivn <int>        max bucket sz as divisor of ref len (default: 4)\n",
      "    --dcv <int>             diff-cover period for blockwise (default: 1024)\n",
      "    --nodc                  disable diff-cover (algorithm becomes quadratic)\n",
      "    -r/--noref              don't build .3/.4 index files\n",
      "    -3/--justref            just build .3/.4 index files\n",
      "    -o/--offrate <int>      SA is sampled every 2^<int> BWT chars (default: 5)\n",
      "    -t/--ftabchars <int>    # of chars consumed in initial lookup (default: 10)\n",
      "    --threads <int>         # of threads\n",
      "    --seed <int>            seed for random number generator\n",
      "    -q/--quiet              verbose output (for debugging)\n",
      "    -h/--help               print detailed description of tool and its options\n",
      "    --usage                 print this usage message\n",
      "    --version               print version information and quit\n"
     ]
    }
   ],
   "source": [
    "# Check the options\n",
    "bowtie2-build --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"bt2_index/sacCer3.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  sacCer3_genome/genome.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 3039276\n",
      "Using parameters --bmax 2279457 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2279457 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1.21571e+07 (target: 2279456)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 12157105 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:03\n",
      "Returning block of 12157106 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3766349\n",
      "fchr[G]: 6086925\n",
      "fchr[T]: 8404025\n",
      "fchr[$]: 12157105\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 8247244 bytes to primary EBWT file: bt2_index/sacCer3.1.bt2\n",
      "Wrote 3039284 bytes to secondary EBWT file: bt2_index/sacCer3.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 12157105\n",
      "    bwtLen: 12157106\n",
      "    sz: 3039277\n",
      "    bwtSz: 3039277\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 759820\n",
      "    offsSz: 3039280\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 63319\n",
      "    numLines: 63319\n",
      "    ebwtTotLen: 4052416\n",
      "    ebwtTotSz: 4052416\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:04\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 3039276\n",
      "Using parameters --bmax 2279457 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 2279457 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 1.21571e+07 (target: 2279456)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 12157105 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:02\n",
      "Returning block of 12157106 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 3766349\n",
      "fchr[G]: 6086925\n",
      "fchr[T]: 8404025\n",
      "fchr[$]: 12157105\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 8247244 bytes to primary EBWT file: bt2_index/sacCer3.rev.1.bt2\n",
      "Wrote 3039284 bytes to secondary EBWT file: bt2_index/sacCer3.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 12157105\n",
      "    bwtLen: 12157106\n",
      "    sz: 3039277\n",
      "    bwtSz: 3039277\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 759820\n",
      "    offsSz: 3039280\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 63319\n",
      "    numLines: 63319\n",
      "    ebwtTotLen: 4052416\n",
      "    ebwtTotSz: 4052416\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:04\n"
     ]
    }
   ],
   "source": [
    "# Create a folder for the Bowtie2 index\n",
    "mkdir -p bt2_index\n",
    "\n",
    "# Create the Bowtie2 index for sacCer3 genome\n",
    "bowtie2-build sacCer3_genome/genome.fa bt2_index/sacCer3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 51040\n",
      "-rw-r--r--  1 cherejirv  1360859114   7.9M Nov 22 16:07 sacCer3.1.bt2\n",
      "-rw-r--r--  1 cherejirv  1360859114   2.9M Nov 22 16:07 sacCer3.2.bt2\n",
      "-rw-r--r--  1 cherejirv  1360859114   161B Nov 22 16:07 sacCer3.3.bt2\n",
      "-rw-r--r--  1 cherejirv  1360859114   2.9M Nov 22 16:07 sacCer3.4.bt2\n",
      "-rw-r--r--  1 cherejirv  1360859114   7.9M Nov 22 16:07 sacCer3.rev.1.bt2\n",
      "-rw-r--r--  1 cherejirv  1360859114   2.9M Nov 22 16:07 sacCer3.rev.2.bt2\n"
     ]
    }
   ],
   "source": [
    "# Let's check the index files that we created\n",
    "ls -lh bt2_index/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many Bowtie2 indexes can be downloaded directly from Illumina\n",
    "https://support.illumina.com/sequencing/sequencing_software/igenome.html\n",
    "\n",
    "For example, the sacCer3 archive is available at\n",
    "ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Saccharomyces_cerevisiae/UCSC/sacCer3/Saccharomyces_cerevisiae_UCSC_sacCer3.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can go ahead and align the short reads to the yeast genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a folder for the aligned data (sam/bam files)\n",
    "mkdir -p bam_files\n",
    "\n",
    "# Align reads with bowtie2, discarding unpaired reads, \n",
    "# reads that are aligned discordantly, and unaligned reads\n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -x bt2_index/sacCer3 \\\n",
    "        -1 fastq_files/test_R1.fastq -2 fastq_files/test_R2.fastq \\\n",
    "        -S bam_files/test.sam 2> bam_files/test.bt2.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 80368\n",
      "-rw-r--r--  1 cherejirv  1360859114   246B Nov 22 16:07 test.bt2.log\n",
      "-rw-r--r--  1 cherejirv  1360859114    39M Nov 22 16:07 test.sam\n"
     ]
    }
   ],
   "source": [
    "# Check the size of the aligned reads (sam file) \n",
    "ls -lh bam_files/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100000 reads; of these:\n",
      "  100000 (100.00%) were paired; of these:\n",
      "    8485 (8.48%) aligned concordantly 0 times\n",
      "    76533 (76.53%) aligned concordantly exactly 1 time\n",
      "    14982 (14.98%) aligned concordantly >1 times\n",
      "91.52% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# View the alignment report generated by bowtie2\n",
    "cat bam_files/test.bt2.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.0\tSO:unsorted\n",
      "@SQ\tSN:chrI\tLN:230218\n",
      "@SQ\tSN:chrII\tLN:813184\n",
      "@SQ\tSN:chrIII\tLN:316620\n",
      "@SQ\tSN:chrIV\tLN:1531933\n",
      "@SQ\tSN:chrIX\tLN:439888\n",
      "@SQ\tSN:chrM\tLN:85779\n",
      "@SQ\tSN:chrV\tLN:576874\n",
      "@SQ\tSN:chrVI\tLN:270161\n",
      "@SQ\tSN:chrVII\tLN:1090940\n",
      "@SQ\tSN:chrVIII\tLN:562643\n",
      "@SQ\tSN:chrX\tLN:745751\n",
      "@SQ\tSN:chrXI\tLN:666816\n",
      "@SQ\tSN:chrXII\tLN:1078177\n",
      "@SQ\tSN:chrXIII\tLN:924431\n",
      "@SQ\tSN:chrXIV\tLN:784333\n",
      "@SQ\tSN:chrXV\tLN:1091291\n",
      "@SQ\tSN:chrXVI\tLN:948066\n",
      "@PG\tID:bowtie2\tPN:bowtie2\tVN:2.3.4.3\tCL:\"/usr/local/bin/../Cellar/bowtie2/2.3.4.3/bin/bowtie2-align-s --wrapper basic-0 --no-mixed --no-discordant -x bt2_index/sacCer3 --passthrough -1 fastq_files/test_R1.fastq -2 fastq_files/test_R2.fastq\"\n",
      "SRR3649298.1\t99\tchrXII\t166119\t42\t50M\t=\t166163\t94\tNTAAACTGTGCTCGGCCAGTGGAGACGTACAAATTCCACTACTATTCGTA\t#<BBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:0T49\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.1\t147\tchrXII\t166163\t42\t50M\t=\t166119\t-94\tTTCGTACTTTGTCTAGTTCAAGGGTCTCCGTCGAGTGAGTCCTACGTTGA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.2\t83\tchrIV\t324159\t42\t50M\t=\t324054\t-155\tCCACAGGTTTTAATAGCCATGTTTTTCCAATTTTTAGAACACAGGGAACN\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBB<#\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:49A0\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.2\t163\tchrIV\t324054\t42\t50M\t=\t324159\t155\tTGCACTGTCAATAAATCATTGGAACTGAATGCTTTGTTGCATGTAGGGCA\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.3\t83\tchrVIII\t113027\t42\t50M\t=\t112907\t-170\tTCAACATTGATTTTTTTTTGTTTCTTTGTAGTTATTAAATGATTTGAGAN\tFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFFBB<<#\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:49A0\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.3\t163\tchrVIII\t112907\t42\t50M\t=\t113027\t170\tTGGCTTGAGGAGTGGTCGTACTGTTGGTCCACCTCACTAACGCAATCATT\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF/BFFBFFF#\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.4\t99\tchrIV\t279819\t42\t50M\t=\t279928\t159\tTCGTTGTGAAATCCATTAAAATTAACACAGGCTCTCAAAAAGGAGGCTTG\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.4\t147\tchrIV\t279928\t42\t50M\t=\t279819\t-159\tTGAAGGCCCTCTTTGTTCAAACTGTCTAGCAAGGTCTGGAGAATTATACA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.5\t99\tchrIX\t408131\t42\t50M\t=\t408231\t150\tNTGTGTAATTGAAGATCCAGGATGTTTTCCTTTTCAGGGAGATGAGAAGG\t#<<BBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:0A49\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.5\t147\tchrIX\t408231\t42\t50M\t=\t408131\t-150\tAAAAATTGTGTCCAATGATTAGCATAGAGAGGTAGAGTATCAGAGAAACA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.6\t99\tchrIII\t46475\t42\t50M\t=\t46597\t172\tTCGTTTAAATATGAGGATTGAGGTAAAATGGTCGTAGAAGGGGCATTATC\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:0\tYT:Z:CP\n"
     ]
    }
   ],
   "source": [
    "# Let's check the first lines of the sam file containing the alignments\n",
    "head -n 30 bam_files/test.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Faster way of running Bowtie2 (use multiple CPUs/threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n"
     ]
    }
   ],
   "source": [
    "# Get the number of CPU cores\n",
    "CORES=$(getconf _NPROCESSORS_ONLN)\n",
    "echo $CORES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run bowtie2 on multiple cores to improve speed\n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -p $((CORES-1)) \\\n",
    "        -x bt2_index/sacCer3 \\\n",
    "        -1 fastq_files/test_R1.fastq -2 fastq_files/test_R2.fastq \\\n",
    "        -S bam_files/test.sam 2> bam_files/test.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OK, that was much faster. Now let's store the alignments in a more compact way (using a `bam` file instead of a `sam` file). We can convert a `sam` file to the `bam` format using `samtools view`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Usage: samtools view [options] <in.bam>|<in.sam>|<in.cram> [region ...]\n",
      "\n",
      "Options:\n",
      "  -b       output BAM\n",
      "  -C       output CRAM (requires -T)\n",
      "  -1       use fast BAM compression (implies -b)\n",
      "  -u       uncompressed BAM output (implies -b)\n",
      "  -h       include header in SAM output\n",
      "  -H       print SAM header only (no alignments)\n",
      "  -c       print only the count of matching records\n",
      "  -o FILE  output file name [stdout]\n",
      "  -U FILE  output reads not selected by filters to FILE [null]\n",
      "  -t FILE  FILE listing reference names and lengths (see long help) [null]\n",
      "  -L FILE  only include reads overlapping this BED FILE [null]\n",
      "  -r STR   only include reads in read group STR [null]\n",
      "  -R FILE  only include reads with read group listed in FILE [null]\n",
      "  -q INT   only include reads with mapping quality >= INT [0]\n",
      "  -l STR   only include reads in library STR [null]\n",
      "  -m INT   only include reads with number of CIGAR operations consuming\n",
      "           query sequence >= INT [0]\n",
      "  -f INT   only include reads with all  of the FLAGs in INT present [0]\n",
      "  -F INT   only include reads with none of the FLAGS in INT present [0]\n",
      "  -G INT   only EXCLUDE reads with all  of the FLAGs in INT present [0]\n",
      "  -s FLOAT subsample reads (given INT.FRAC option value, 0.FRAC is the\n",
      "           fraction of templates/read pairs to keep; INT part sets seed)\n",
      "  -M       use the multi-region iterator (increases the speed, removes\n",
      "           duplicates and outputs the reads as they are ordered in the file)\n",
      "  -x STR   read tag to strip (repeatable) [null]\n",
      "  -B       collapse the backward CIGAR operation\n",
      "  -?       print long help, including note about region specification\n",
      "  -S       ignored (input format is auto-detected)\n",
      "      --input-fmt-option OPT[=VAL]\n",
      "               Specify a single input file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
      "               Specify output format (SAM, BAM, CRAM)\n",
      "      --output-fmt-option OPT[=VAL]\n",
      "               Specify a single output file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -T, --reference FILE\n",
      "               Reference sequence FASTA FILE [null]\n",
      "  -@, --threads INT\n",
      "               Number of additional threads to use [0]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Check the arguments of samtools view\n",
    "samtools view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the sam file to bam format\n",
    "samtools view -b -h --threads $((CORES-1)) bam_files/test.sam > bam_files/test.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 94328\n",
      "-rw-r--r--  1 cherejirv  1360859114   7.0M Nov 22 16:07 test.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114   246B Nov 22 16:07 test.bt2.log\n",
      "-rw-r--r--  1 cherejirv  1360859114    39M Nov 22 16:07 test.sam\n"
     ]
    }
   ],
   "source": [
    "# Let's compare the sizes of the sam and bam files\n",
    "ls -lh bam_files/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `bam` file is about 5 times smaller than the corresponding `sam` file. `Bam` is a binary file format so trying to read it as a regular text won't work.\n",
    "```\n",
    "head bam/test.bam\n",
    "```\n",
    "will output some nonsense. The proper way of viewing a `bam` file is using `samtools view`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3649298.1\t99\tchrXII\t166119\t42\t50M\t=\t166163\t94\tNTAAACTGTGCTCGGCCAGTGGAGACGTACAAATTCCACTACTATTCGTA\t#<BBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:0T49\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.1\t147\tchrXII\t166163\t42\t50M\t=\t166119\t-94\tTTCGTACTTTGTCTAGTTCAAGGGTCTCCGTCGAGTGAGTCCTACGTTGA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.2\t83\tchrIV\t324159\t42\t50M\t=\t324054\t-155\tCCACAGGTTTTAATAGCCATGTTTTTCCAATTTTTAGAACACAGGGAACN\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBB<#\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:49A0\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.2\t163\tchrIV\t324054\t42\t50M\t=\t324159\t155\tTGCACTGTCAATAAATCATTGGAACTGAATGCTTTGTTGCATGTAGGGCA\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.3\t83\tchrVIII\t113027\t42\t50M\t=\t112907\t-170\tTCAACATTGATTTTTTTTTGTTTCTTTGTAGTTATTAAATGATTTGAGAN\tFFFFFFFFFFBFFFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFFBB<<#\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:49A0\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.3\t163\tchrVIII\t112907\t42\t50M\t=\t113027\t170\tTGGCTTGAGGAGTGGTCGTACTGTTGGTCCACCTCACTAACGCAATCATT\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF/BFFBFFF#\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n",
      "SRR3649298.4\t99\tchrIV\t279819\t42\t50M\t=\t279928\t159\tTCGTTGTGAAATCCATTAAAATTAACACAGGCTCTCAAAAAGGAGGCTTG\tBBBBBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.4\t147\tchrIV\t279928\t42\t50M\t=\t279819\t-159\tTGAAGGCCCTCTTTGTTCAAACTGTCTAGCAAGGTCTGGAGAATTATACA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.5\t99\tchrIX\t408131\t42\t50M\t=\t408231\t150\tNTGTGTAATTGAAGATCCAGGATGTTTTCCTTTTCAGGGAGATGAGAAGG\t#<<BBFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:-1\tXN:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tNM:i:1\tMD:Z:0A49\tYS:i:0\tYT:Z:CP\n",
      "SRR3649298.5\t147\tchrIX\t408231\t42\t50M\t=\t408131\t-150\tAAAAATTGTGTCCAATGATTAGCATAGAGAGGTAGAGTATCAGAGAAACA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBBBBB\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:50\tYS:i:-1\tYT:Z:CP\n"
     ]
    }
   ],
   "source": [
    "# Proper way of listing the alignments from a bam file \n",
    "# (use -h to include the header information)\n",
    "samtools view bam_files/test.bam | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  183030\n"
     ]
    }
   ],
   "source": [
    "# Count the total number of alignments\n",
    "samtools view bam_files/test.bam | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 183,030 reads, corresponding to 91,515 DNA fragments (paired-end reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100000 reads; of these:\n",
      "  100000 (100.00%) were paired; of these:\n",
      "    8485 (8.48%) aligned concordantly 0 times\n",
      "    76533 (76.53%) aligned concordantly exactly 1 time\n",
      "    14982 (14.98%) aligned concordantly >1 times\n",
      "91.52% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "# Let's look again at the bowtie2 log:\n",
    "cat bam_files/test.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These 91,515 properly aligned pairs are the sum of: 76,533 + 14,982"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore the bam file in more detail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3649298.1\t99\tchrXII\t166119\t42\n",
      "SRR3649298.1\t147\tchrXII\t166163\t42\n",
      "SRR3649298.2\t83\tchrIV\t324159\t42\n",
      "SRR3649298.2\t163\tchrIV\t324054\t42\n",
      "SRR3649298.3\t83\tchrVIII\t113027\t42\n",
      "SRR3649298.3\t163\tchrVIII\t112907\t42\n",
      "SRR3649298.4\t99\tchrIV\t279819\t42\n",
      "SRR3649298.4\t147\tchrIV\t279928\t42\n",
      "SRR3649298.5\t99\tchrIX\t408131\t42\n",
      "SRR3649298.5\t147\tchrIX\t408231\t42\n"
     ]
    }
   ],
   "source": [
    "# Show the first 5 columns of the bam file\n",
    "samtools view bam_files/test.bam | head | cut -f 1-5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first 5 columns represent the following quantities:  \n",
    "Column 1: alignment name.  \n",
    "Column 2: sum of multiple bitwise flags. See https://samtools.github.io/hts-specs/SAMv1.pdf for an explanation of different FLAGs. One can easily check the meaning of each FLAG using the following website:\n",
    "https://broadinstitute.github.io/picard/explain-flags.html  \n",
    "Column 3: chromosome name.  \n",
    "Column 4: mapping position of the left end of the read.  \n",
    "Column 5: mapping quality, i.e. $−10 \\log_{10}\\times$ Probability that the mapping position is wrong."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50M\t=\t166163\t94\n",
      "50M\t=\t166119\t-94\n",
      "50M\t=\t324054\t-155\n",
      "50M\t=\t324159\t155\n",
      "50M\t=\t112907\t-170\n",
      "50M\t=\t113027\t170\n",
      "50M\t=\t279928\t159\n",
      "50M\t=\t279819\t-159\n",
      "50M\t=\t408231\t150\n",
      "50M\t=\t408131\t-150\n"
     ]
    }
   ],
   "source": [
    "# Show columns 6 to 9\n",
    "samtools view bam_files/test.bam | head | cut -f 6-9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These columns represent the following:  \n",
    "Column 6: CIGAR string  \n",
    "Column 7: ‘=’ if the paired read was aligned on the same chromosome  \n",
    "Column 8: position of the paired read  \n",
    "Column 9: length of the whole DNA fragment whose ends were sequenced (+/- for left/right end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NTAAACTGTGCTCGGCCAGTGGAGACGTACAAATTCCACTACTATTCGTA\n",
      "TTCGTACTTTGTCTAGTTCAAGGGTCTCCGTCGAGTGAGTCCTACGTTGA\n",
      "CCACAGGTTTTAATAGCCATGTTTTTCCAATTTTTAGAACACAGGGAACN\n",
      "TGCACTGTCAATAAATCATTGGAACTGAATGCTTTGTTGCATGTAGGGCA\n",
      "TCAACATTGATTTTTTTTTGTTTCTTTGTAGTTATTAAATGATTTGAGAN\n",
      "TGGCTTGAGGAGTGGTCGTACTGTTGGTCCACCTCACTAACGCAATCATT\n",
      "TCGTTGTGAAATCCATTAAAATTAACACAGGCTCTCAAAAAGGAGGCTTG\n",
      "TGAAGGCCCTCTTTGTTCAAACTGTCTAGCAAGGTCTGGAGAATTATACA\n",
      "NTGTGTAATTGAAGATCCAGGATGTTTTCCTTTTCAGGGAGATGAGAAGG\n",
      "AAAAATTGTGTCCAATGATTAGCATAGAGAGGTAGAGTATCAGAGAAACA\n"
     ]
    }
   ],
   "source": [
    "# Show column 10: DNA sequence\n",
    "samtools view bam_files/test.bam | head | cut -f 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default the sam/bam files are sorted by the first column (read name)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3649298.1\tchrXII\t166119\n",
      "SRR3649298.1\tchrXII\t166163\n",
      "SRR3649298.2\tchrIV\t324159\n",
      "SRR3649298.2\tchrIV\t324054\n",
      "SRR3649298.3\tchrVIII\t113027\n",
      "SRR3649298.3\tchrVIII\t112907\n",
      "SRR3649298.4\tchrIV\t279819\n",
      "SRR3649298.4\tchrIV\t279928\n",
      "SRR3649298.5\tchrIX\t408131\n",
      "SRR3649298.5\tchrIX\t408231\n"
     ]
    }
   ],
   "source": [
    "# Show the following 3 columns: 1 - read name, 3 - chromosome, 4 - position)\n",
    "samtools view bam_files/test.bam | head | cut -f 1,3,4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many tools require the bam file to be sorted such that the alignments occur in “genome order”. That is, ordered positionally based upon their alignment coordinates on each chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: samtools sort [options...] [in.bam]\n",
      "Options:\n",
      "  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)\n",
      "  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]\n",
      "  -n         Sort by read name\n",
      "  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)\n",
      "  -o FILE    Write final output to FILE rather than standard output\n",
      "  -T PREFIX  Write temporary files to PREFIX.nnnn.bam\n",
      "      --input-fmt-option OPT[=VAL]\n",
      "               Specify a single input file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
      "               Specify output format (SAM, BAM, CRAM)\n",
      "      --output-fmt-option OPT[=VAL]\n",
      "               Specify a single output file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "      --reference FILE\n",
      "               Reference sequence FASTA FILE [null]\n",
      "  -@, --threads INT\n",
      "               Number of additional threads to use [0]\n"
     ]
    }
   ],
   "source": [
    "# We can re-sort the BAM file using \n",
    "samtools sort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 11 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "samtools sort -o bam_files/test.sorted.bam --threads $((CORES-1)) bam_files/test.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3649298.54916\tchrI\t231\n",
      "SRR3649298.54916\tchrI\t293\n",
      "SRR3649298.36570\tchrI\t301\n",
      "SRR3649298.26155\tchrI\t309\n",
      "SRR3649298.36570\tchrI\t409\n",
      "SRR3649298.26155\tchrI\t410\n",
      "SRR3649298.18919\tchrI\t415\n",
      "SRR3649298.18919\tchrI\t433\n",
      "SRR3649298.60307\tchrI\t854\n",
      "SRR3649298.60307\tchrI\t918\n"
     ]
    }
   ],
   "source": [
    "# Check the sorted bam file (columns 1 - read name, 3 - chromosome, 4 - position)\n",
    "samtools view bam_files/test.sorted.bam | head | cut -f 1,3,4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: samtools index [-bc] [-m INT] <in.bam> [out.index]\n",
      "Options:\n",
      "  -b       Generate BAI-format index for BAM files [default]\n",
      "  -c       Generate CSI-format index for BAM files\n",
      "  -m INT   Set minimum interval size for CSI indices to 2^INT [14]\n",
      "  -@ INT   Sets the number of threads [none]\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Indexing a position sorted bam file allows one to quickly extract alignments overlapping \n",
    "# particular genomic regions. The indexing is done using\n",
    "samtools index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools index -@ $((CORES-1)) bam_files/test.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 106672\n",
      "-rw-r--r--  1 cherejirv  1360859114   7.0M Nov 22 16:07 test.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114   246B Nov 22 16:07 test.bt2.log\n",
      "-rw-r--r--  1 cherejirv  1360859114    39M Nov 22 16:07 test.sam\n",
      "-rw-r--r--  1 cherejirv  1360859114   6.0M Nov 22 16:07 test.sorted.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114   9.5K Nov 22 17:09 test.sorted.bam.bai\n"
     ]
    }
   ],
   "source": [
    "# Check the sizes (the index file .bai is very small)\n",
    "ls -lh bam_files/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pipes\n",
    "All the above operations could be run in a more compact way, using pipes (see https://www.geeksforgeeks.org/piping-in-unix-or-linux/ for more info on the pipe operator). Let's align now all the reads (not only 100,000 reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 11 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "# Using pipes to perform all these operations in one step\n",
    "bowtie2 --no-mixed --no-discordant --no-unal \\\n",
    "        -p $((CORES-1)) \\\n",
    "        -x bt2_index/sacCer3 \\\n",
    "        -1 fastq_files/SRR3649298_1.fastq.gz \\\n",
    "        -2 fastq_files/SRR3649298_2.fastq.gz \\\n",
    "        2> bam_files/SRR3649298.bt2.log \\\n",
    "        | samtools view -b -@ $((CORES-1)) - \\\n",
    "        | samtools sort -o bam_files/SRR3649298.sorted.bam -@ $((CORES-1)) -\n",
    "samtools index -@ $((CORES-1)) bam_files/SRR3649298.sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 963936\n",
      "-rw-r--r--  1 cherejirv  1360859114   256B Nov 22 17:15 SRR3649298.bt2.log\n",
      "-rw-r--r--  1 cherejirv  1360859114   414M Nov 22 17:15 SRR3649298.sorted.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114    37K Nov 22 17:15 SRR3649298.sorted.bam.bai\n",
      "-rw-r--r--  1 cherejirv  1360859114   7.0M Nov 22 16:07 test.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114   246B Nov 22 16:07 test.bt2.log\n",
      "-rw-r--r--  1 cherejirv  1360859114    39M Nov 22 16:07 test.sam\n",
      "-rw-r--r--  1 cherejirv  1360859114   6.0M Nov 22 16:07 test.sorted.bam\n",
      "-rw-r--r--  1 cherejirv  1360859114   9.5K Nov 22 17:09 test.sorted.bam.bai\n"
     ]
    }
   ],
   "source": [
    "ls -lh bam_files/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11358904 reads; of these:\n",
      "  11358904 (100.00%) were paired; of these:\n",
      "    990760 (8.72%) aligned concordantly 0 times\n",
      "    8670333 (76.33%) aligned concordantly exactly 1 time\n",
      "    1697811 (14.95%) aligned concordantly >1 times\n",
      "91.28% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "cat bam_files/SRR3649298.bt2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze the aligned data  \n",
    "We'll switch now to another Jupyter notebook and analyze the aligned data in R.\n",
    "\n",
    "### Further reading resources:  \n",
    "https://samtools.github.io/hts-specs/SAMv1.pdf  \n",
    "http://quinlanlab.org/tutorials/samtools/samtools.html  \n",
    "http://biobits.org/samtools_primer.html"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
